vnext/Shared/JSI/NapiJsiV8RuntimeHolder.cpp

```diff
@@ -0,0 +1,120 @@
+      : m_env{env},
+        m_taskCallback{taskCallback},
+        m_taskData{taskData},
+        m_finalizeCallback{finalizeCallback},
+        m_finalizeHint{finalizeHint} {}
+
+  NapiTask(const NapiTask &) = delete;
+  NapiTask &operator=(const NapiTask &) = delete;
+
+  ~NapiTask() {
+    if (m_finalizeCallback) {
+      m_finalizeCallback(m_env, m_taskData, m_finalizeHint);
+    }
+  }
+
+  void operator()() noexcept {
+    m_taskCallback(m_env, m_taskData);
+  }
+
+ private:
+  napi_env m_env;
+  napi_ext_task_callback m_taskCallback;
+  void *m_taskData;
+  napi_finalize m_finalizeCallback;
+  void *m_finalizeHint;
+};
+
+// See napi_ext_schedule_task_callback definition.
+/*static*/ void NapiJsiV8RuntimeHolder::ScheduleTaskCallback(
+    napi_env env,
+    napi_ext_task_callback taskCallback,
+    void *taskData,
+    uint32_t /*delayInMsec*/,
+    napi_finalize finalizeCallback,
+    void *finalizeHint) {
+  NapiJsiV8RuntimeHolder *holder;
+  auto result = napi_get_instance_data(env, (void **)&holder);
+  if (result != napi_status::napi_ok) {
+    std::terminate();
+  }
+
+  auto task = std::make_shared<NapiTask>(env, taskCallback, taskData, finalizeCallback, finalizeHint);
+  holder->m_jsQueue->runOnQueue([task = std::move(task)]() { task->operator()(); });
+}
+
+NapiJsiV8RuntimeHolder::NapiJsiV8RuntimeHolder(
+    shared_ptr<DevSettings> devSettings,
+    shared_ptr<MessageQueueThread> jsQueue,
+    unique_ptr<ScriptStore> &&scriptStore,
+    unique_ptr<PreparedScriptStore> &&preparedScritpStore) noexcept
+    : m_useDirectDebugger{devSettings->useDirectDebugger},
+      m_debuggerBreakOnNextLine{devSettings->debuggerBreakOnNextLine},
+      m_debuggerPort{devSettings->debuggerPort},
+      m_debuggerRuntimeName{devSettings->debuggerRuntimeName},
+      m_jsQueue{std::move(jsQueue)},
+      m_scriptStore{std::move(scriptStore)},
+      m_preparedScriptStore{std::move(preparedScritpStore)} {}
+
+void NapiJsiV8RuntimeHolder::InitRuntime() noexcept {
+  napi_env env{};
+  napi_ext_env_settings settings{};
+  settings.this_size = sizeof(napi_ext_env_settings);
+  settings.flags.enable_gc_api = true;
+  if (m_debuggerPort > 0)
+    settings.inspector_port = m_debuggerPort;
+
+  settings.flags.enable_inspector = m_useDirectDebugger;
+  settings.flags.wait_for_debugger = m_debuggerBreakOnNextLine;
+  // TODO: debuggerRuntimeName?
+
+  settings.foreground_scheduler = &NapiJsiV8RuntimeHolder::ScheduleTaskCallback;
+  // TODO: scriptStore?
+
+  napi_ext_create_env(&settings, &env);
+  // Associate environment to holder.
+  napi_set_instance_data(env, this, nullptr, nullptr); // TODO: Finalize
+
+  m_runtime = MakeNapiJsiRuntime(env);
+  m_ownThreadId = std::this_thread::get_id();
+}
+
+#pragma region facebook::jsi::RuntimeHolderLazyInit
+
+shared_ptr<Runtime> NapiJsiV8RuntimeHolder::getRuntime() noexcept /*override*/
+{
+  std::call_once(m_onceFlag, [this]() { InitRuntime(); });
+
+  if (!m_runtime)
+    std::terminate();
```

Comment:
seems wrong to terminate like this. can we make this non-noexcept and throw?

Reply:
This method implements a function pointer spec defined in V8/NAPI. It is already non-noexcept because it's meant to behave like a plain C function.
We can't throw because it is meant to be ABI-safe and exceptions behave differently in say, MSVC and LLVM-produced runtimes. @vmoroz can you please confirm?

`std::termintate` was the accepted solution for now after offline discussion.

Reply:
was there a design discussion on all of this? if not, let's have one please.

Reply:
@NikoAri @tudorms @vmoroz 

Given that the callback this free function implements must be ABI-safe (non-throwing), is `std::terminate` an acceptable measure or do we need to change V8JSI to use an error-signaling mechanism?

Reply:
nit: at a minimum, we should probably log for local dev, and ensure we get a watson report, see: https://github.com/microsoft/react-native-windows/issues/3454. We have a bunch of other std:terminate calls in our codebase. If we decide to remain std:terminate I'd love us to consider creating a shared helper function for these scenario's that take care of the logging and watson report.

Reply:
can we call errorCallback() here and let users decide whether to crash or not?

Reply:
> can we call errorCallback() here and let users decide whether to crash or not?

We cannot. The runtime holder does not hold a reference to the `DevSettings` object 
The current design for V8 runtime holders uses this mechanism.
Changing this would likely require changing V8JSI itself, which is out of the scope of this PR.

Reply:
> nit: at a minimum, we should probably log for local dev, and ensure we get a watson report, see: #3454. We have a bunch of other std:terminate calls in our codebase. If we decide to remain std:terminate I'd love us to consider creating a shared helper function for these scenario's that take care of the logging and watson report.

Agreed, but this would be out of the scope of this PR.
I'd rather not make it a spanning change, especially because it needs to be backported ASAP, and it does not introduce this issue.